src/ui/public/embeddable/types.ts

```diff
@@ -0,0 +1,57 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+export interface ContainerState {
+  // 'view' or 'edit'. Should probably be an enum but I'm undecided where to define it, here or in dashboard code.
+  viewMode: string;
+
+  timeRange: {
+    // To and From should be either an absolute time range in utc format or a relative one (e.g. now-15m)
+    to: string;
+    from: string;
+  };
+
+  // The shape will be up to the embeddable type.
+  embeddableCustomization?: object;
+
+  /**
+   * Whether or not panel titles are hidden. It is not the embeddable's responsibility to hide the title (the container
+   * handles that). This information is currently only used to determine the title for reporting (data-sharing-title
+   * attribute). If we move that out of the embeddables and push it to the container (as we probably should), then
+   * we shouldn't need to expose this information.
+   */
+  hidePanelTitles: boolean;
+
+  /**
+   * Is the current panel in expanded mode
+   */
+  isPanelExpanded: boolean;
+}
+
+export interface EmbeddableState {
+  /**
+   * Any customization data that should be stored at the panel level. For
+   * example, pie slice colors, or custom per panel sort order or columns.
+   */
+  customization: object;
```

Comment:
__note:__ hopefully we'll have more detailed types for these at some point :)

Reply:
Actually this will remain an object because it's a place for the plug in to store whatever it wants. We shouldn't make any assumptions about what's inside.  When we get more specific is inside visualize embeddable and saved search embedable. Each of their customization objects look totally different.  

I suppose if we required a bit more of a consistent format (maybe variable definitions - like variable name, input type, value, possible values) we could expose these variables and make them editable in a panel.  But that would be long term.  For now, each embeddable's customization object can be totally unique.

Reply:
Hmm, then type of `customization` sounds like a generic type parameter `EmbeddableState<TCustomization extends object> { customization: TCustomization; }`, but agree it's too early to do something here. We can get back to it once we have more TypeScript consumers of `EmbeddableState` to see how we can improve that.

Reply:
Ahh, yes that would be nice! Wow it's been so long since I've written in a typed language I've forgotten all these awesome features. Can't wait till more code is typescriptified to take advantage of all of it.